cleanup.go

```diff
@@ -43,15 +49,44 @@ func (s *CleanupSuite) callStack(c *gc.C, stack cleanupStack) {
 }
 
 // AddCleanup pushes the cleanup function onto the stack of functions to be
-// called during TearDownTest.
+// called during TearDownTest or TearDownSuite. TearDownTest will be used if
+// SetUpTest has already been called, else we will use TearDownSuite
 func (s *CleanupSuite) AddCleanup(cleanup CleanupFunc) {
+	if s.suiteSuite == nil {
+		// This is either called before SetUpSuite or after
+		// TearDownSuite. Either way, we can't really trust that we're
+		// going to call Cleanup correctly.
+		panic("unsafe to call AddCleanup without a Suite")
+	}
+	if s != s.suiteSuite {
+		// If you write a test like:
+		// func (s MySuite) TestFoo(c *gc.C) {
+		//   s.AddCleanup(foo)
+		// }
+		// The AddCleanup call is unsafe because it modifes
+		// s.suiteSuite but that object disappears once TestFoo
+		// returns. So you have to use:
+		// func (s *MySuite) TestFoo(c *gc.C) if you want the Cleanup
+		// funcs.
+		panic("unsafe to call AddCleanup from non pointer receiver test")
+	}
+	if s.testSuite == nil {
+		// We either haven't called SetUpTest or we've already called
+		// TearDownTest, consider this a Suite level cleanup.
+		s.suiteStack = append(s.suiteStack, cleanup)
+		return
+	}
+	if s != s.testSuite {
```

Comment:
Confirmed. I removed it and the test suite still passed which should catch this sort of thing.